lib/upload.js

```diff
@@ -0,0 +1,78 @@
+"use strict";
+
+module.exports = (fsAsync, log, morePromises, now, path, randomstring, shelfError) => {
+    const DEDUPLICATE_ERROR_CODE_LIST = [
+        shelfError.UNAUTHORIZED,
+        shelfError.NOT_FOUND
+    ];
+
+    /**
+     * This function exists in order to map a few errors so that the error
+     * messages are more readable. It will also attempt to deduplicate some
+     * errors that likely happened all for the same reason. For example, if
+     * you uploaded 10 artifacts and your token was invalid, you probably would
+     * not want to be told so 10 times.
+     *
+     * @param {Error[]} errorList
+     * @param {shelfLib.Reference} reference
+     * @return {Array.<string>}
```

Comment:
Looks like you do both `Error[]` and `Array.<string>` which are both valid but pointing out for consistency I guess. don't care about it too much just noticed it.